src/Utilities/Iterator.f90

```diff
@@ -0,0 +1,37 @@
+module IteratorModule
+
+  implicit none
+  private
+
+  public :: IteratorType
+
+  type, abstract :: IteratorType
+  contains
+    procedure(has_next_if), deferred :: has_next
+    procedure(next_if), deferred :: next
+    procedure(value_if), deferred :: value
+
+  end type IteratorType
+
+  abstract interface
+
+    function has_next_if(this) result(res)
+      import IteratorType
+      class(IteratorType) :: this
+      type(logical) :: res
```

Comment:
Although debatable, we generally work with kind LGP for logicals, so: logical(LGP) :: ...